### DISCUSSION TITLE
I need help not related to littlenavmap, but my own project i am working on. I have to create my own Pages(Flight plan and so on).

### DISCUSSION THREAD

**Discussion by Mohammad-Imran01** [albar965/littlenavmap#1092] [General]
If i use use qml which is a little easier for me to style and design, it puts alot of load on application.
I dont want to use qml can you please tell me if you were in my situation how would you have designed such UI for example Flight plan UI i am showing below.

![flightPlan](https://github.com/albar965/littlenavmap/assets/117526105/b4c2b437-80c1-47b8-b5ae-8eebb31acb02)

I have to create 20/25 such pages, but just 4 qml pages put alot of pressure. Note: I used map widget as parent

**Comment by albar965** [comment]
Keep using the map widget as parent but put all pages into windows on top of it. Then fill these windows with content using the QPainter and C++. This is a lot of work. 
You can also try to use Qt widgets like QScrollArea or QTableWidget and change the look and feel using stylesheets until they look like the G1000 pages
Not sure if there are better approaches. Javascript and Chrome engine will also consume a lot of resources.

If the QML puts too much strain on the application you should check your code. The fewest tasks have to be run every frame.

**Reply by Mohammad-Imran01** [reply]
Thank you. Now i can use QWidgets confidently.

**Comment by Mohammad-Imran01** [comment]
Hello, I came across Flight gear G1000, which uses exactly same pages as i need but SVG files and dynamically changing values, If you have worked with SVGs for QT GUI, can you please tell me, can i also use SVGs? Will it work fine like Widgets and QML components?

**Reply by albar965** [reply]
You can use SVG files. All icons in LNM toolbars and on the map are SVG.

See https://doc.qt.io/qt-5/svgrendering.html
But keep in mind that
> Qt supports the static features of SVG 1.2 Tiny. ECMA scripts and DOM manipulation are currently not supported.

Alex

**Comment by Mohammad-Imran01** [comment]
I need your help please, 
I am facing compilation error because of flightplanconstants.h(maybe) it randomly comes out of nowhere, note i havent edited any atools files neither i am using flightplans files till now in my classes. But the errors(40, 70 or more than 100) randomly comes and takes alot of time. I have solved it by removing certain headers files, last time i remember i had to remove "//#include "src/route/routecontroller.h", 
Do you know this issue? is there any parmanent solution?

/******   Same file (flightplanconstants.h:) hundreds of errors, without changing or including this file. ********/
C:\Users\Imran\Desktop\Aexr\DA42-G1000\atools\src\fs\pln\flightplanconstants.h:63: error: expected identifier before numeric constant
In file included from D:/Apps/Qt/Tools/mingw810_32/i686-w64-mingw32/include/windows.h:71,
                 from D:\Apps\Qt\5.15.2\mingw81_32\include/QtCore/qt_windows.h:64,
                 from D:\Apps\Qt\5.15.2\mingw81_32\include/QtGui/qopengl.h:49,
                 from D:\Apps\Qt\5.15.2\mingw81_32\include/QtQuick/qsggeometry.h:44,
                 from D:\Apps\Qt\5.15.2\mingw81_32\include/QtQuick/qsgnode.h:43,
                 from D:\Apps\Qt\5.15.2\mingw81_32\include/QtQuick/qsgrendererinterface.h:43,
                 from D:\Apps\Qt\5.15.2\mingw81_32\include/QtQuick/qquickwindow.h:44,
                 from D:\Apps\Qt\5.15.2\mingw81_32\include\QtQuickWidgets/qquickwidget.h:44,
                 from D:\Apps\Qt\5.15.2\mingw81_32\include\QtQuickWidgets/QQuickWidget:1,
                 from ..\littlenavmap/custom/info/pagetextbuilder.h:20,
                 from ..\littlenavmap/custom/pageGroupWidgets/pgwpt.h:4,
                 from ..\littlenavmap\custom\info\pagedatacontroller.h:7,
                 from ..\littlenavmap\custom\info\pagedatacontroller.cpp:1:
..\atools\src/fs/pln/flightplanconstants.h: At global scope:
..\atools\src/fs/pln/flightplanconstants.h:63:3: error: expected identifier before numeric constant
   ALTERNATE = 1 << 2, /* Flight plan entry leads to an alternate airport */
   ^~~~~~~~~


![Screenshot (32)](https://github.com/albar965/littlenavmap/assets/117526105/79d3b8e3-acd7-40b3-a003-9565ddbd90f6)

**Reply by albar965** [reply]
Do you use the matching latest branches? This is `release/4.0` atools and `release/3.0` littlenavmap.

Also your flightplanconstants.h has a problem:
`ALTERNATE = 1 << 2, / Flight plan entry leads to an alternate airport */`
should be 
`ALTERNATE = 1 << 2, /* Flight plan entry leads to an alternate airport */`
Maybe you accidentally delete the `*`.

**Reply by Mohammad-Imran01** [reply]
I am using the same version, i had downloaded 4/5 months ago, everything works fine just including some files like mapflags gives this error. And i dont know why the error shows like that but there is no missing "*" or anything.

You have never faced this issue?

**Reply by albar965** [reply]
Never seen this but your pasted error message shows clearly that the comment is broken.

> error: expected identifier before numeric constant

This also means that the comment is broken. Revert the changes, update from Github and try again. It's just a `*` :slightly_smiling_face:

**Reply by Mohammad-Imran01** [reply]
Sorry for bothering but, see it yourself, the Flag is just above Q_DECLARE_FLAGS (without missing astrisk) still it says undefined giving same error like it did few months(and few weeks) back.



`/*****************************************************************************
* Copyright 2015-2023 Alexander Barthel alex@littlenavmap.org
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program.  If not, see <http://www.gnu.org/licenses/>.
*****************************************************************************/

#ifndef ATOOLS_FLIGHTPLANCONSTANTS_H
#define ATOOLS_FLIGHTPLANCONSTANTS_H

#include <limits>

#include <QLatin1String>

namespace atools {
namespace fs {
namespace pln {

namespace pattern {

/* Pattern variables for flight plan names */
const QLatin1String PLANTYPE("PLANTYPE"); // IFR or VFR
const QLatin1String DEPARTIDENT("DEPARTIDENT"); // Departure airport ICAO code
const QLatin1String DEPARTNAME("DEPARTNAME"); // Departure airport name
const QLatin1String DESTIDENT("DESTIDENT"); // Destination airport ICAO code
const QLatin1String DESTNAME("DESTNAME"); // Destination airport name
const QLatin1String CRUISEALT("CRUISEALT"); // Cruise altitude

/* Default patterns */
const QString SHORT(DEPARTIDENT + " " + DESTIDENT);
const QString LONG(PLANTYPE + " " + DEPARTNAME + " (" + DEPARTIDENT + ") to " + DESTNAME + " (" + DESTIDENT + ")");
}

const static float FLIGHTPLAN_ALTITUDE_FT_MIN = -2000.f;
const static float FLIGHTPLAN_ALTITUDE_FT_MAX = 100000.f; // GUI altitude limit

namespace entry {

enum WaypointType
{
  UNKNOWN,
  AIRPORT,
  WAYPOINT,
  VOR,
  NDB,
  USER
};

enum Flag
{
  NONE = 0,
  PROCEDURE = 1 << 1, /* Flight plan entry is any procedure leg */
  ALTERNATE = 1 << 2, /* Flight plan entry leads to an alternate airport */
  TRACK = 1 << 3 /* Flight plan entry airway name is a track */
};

Q_DECLARE_FLAGS(Flags, Flag);
Q_DECLARE_OPERATORS_FOR_FLAGS(atools::fs::pln::entry::Flags);
}

/* File format for flight plans as detected by FlightplanIO::load and detectFormat */
enum FileFormat
{
  NONE,
  FSX_PLN, /* FSX or P3D XML PLN flight plan - can load and save */
  FS9_PLN, /* FS9 ini style PLN flight plan - can load only */
  FMS3, /* X-Plane version 3 FMS file - can load and save */
  FMS11, /* X-Plane version 11 FMS file - can load and save */
  FLP, /* Aerosoft airbus or FlightFactor Boeing - can load and save */
  FSC_PLN, /* FSC ini style PLN flight plan - can load only */
  FLIGHTGEAR, /* FlightGear XML format - load and save  */
  LNM_PLN, /* Little Navmap's own flight plan format  */
  GARMIN_FPL, /* Garmin FPL - XML format */
  MSFS_PLN, /* MSFS PLN  */
  GARMIN_GFP /* Garmin GFP - One line text format prefixed with "FPN/RI:..." */
};

enum FlightplanType
{
  NO_TYPE,
  IFR,
  VFR
};

/* Departure parking type. Describes content of FlightPlan::departureParkingName */
enum FlightplanParkingType
{
  NO_POS,
  AIRPORT,
  RUNWAY, /* Zero prefixed runway number like "07C" */
  PARKING, /* Free X-Plane parking name or FSX/P3D/MSFS naming scheme. */
  HELIPAD /* Number describing the helipad */
};

/* Currently only used for saving FSX/P3D PLN files */
enum RouteType
{
  LOW_ALTITUDE,
  HIGH_ALTITUDE,
  VOR, /* Used for radio navaid routing (VOR and NDB) */
  DIRECT, /* Direct connection without waypoints */
  UNKNOWN /* Has to be changed later when resolving the ident to database objects */
};

/*
 * Common key that are used int flight plan properties that are not supported in PLN.
 * Used to transport more optional details to the export methods.
 * Keys should never be empty to avoid issues when loading procedures. */

/* Keys that describe procedures */
/* SID, runway and transiton */
const QLatin1String SID("sidappr"); /* SID name */
const QLatin1String SID_RW("sidapprrw"); /* SID runway */
const QLatin1String SID_TRANS("sidtrans"); /* SID transition name */
const QLatin1String SID_TRANS_WP("sidtranswp"); /* Alternative to above, the endpoint of the SID transition */
const QLatin1String SID_TYPE("sidtype"); /* Optional type of SID, CUSTOMDEPART for selected runway */
const QLatin1String SID_TYPE_CUSTOM("CUSTOMDEPART"); /* SIDTYPE for user defined departure */

/* STAR, runway and transiton */
const QLatin1String STAR("star"); /* STAR name */
const QLatin1String STAR_RW("starrw"); /* STAR runway */
const QLatin1String STAR_TRANS("startrans"); /* STAR transition name */
const QLatin1String STAR_TRANS_WP("startranswp"); /* Alternative to above, the possible startpoints of a STAR transition
                                                   *  separated by PROPERTY_LIST_SEP */

/* Approach, runway and more */
const QLatin1String APPROACH("approach"); /* Approach name like waypoint */
const QLatin1String APPROACH_ARINC("approacharinc"); /* ARINC short name for FMS files */
const QLatin1String APPROACH_TYPE("approachtype"); /* Optional type like ILS, RNAV, etc. CUSTOM for selected arrival runway and leg distances. */
const QLatin1String APPROACH_RW("approachrw"); /* Runway for approach */
const QLatin1String APPROACH_SUFFIX("approachsuffix"); /* Suffix like Z for ILS-Z */
const QLatin1String APPROACH_TYPE_CUSTOM("CUSTOM"); /* User defined departure in APPROACHTYPE */

/* Approach transiton */
const QLatin1String TRANSITION("transition"); /* Transition name */
const QLatin1String TRANSITION_TYPE("transitiontype"); /* Type: Full, etc. */

/* Name of airway which leads to IAF of a STAR or an approach (-transition) */
const QLatin1String PROCAIRWAY("procairway");

/* Only for APPROACHTYPE = CUSTOM */
const QLatin1String APPROACH_CUSTOM_DISTANCE("approachcustomdistance"); /* Length or final leg in NM */
const QLatin1String APPROACH_CUSTOM_ALTITUDE("approachcustomaltitude"); /* Entry altitude at final leg in ft */
const QLatin1String APPROACH_CUSTOM_OFFSET("approachcustomoffset"); /* Offset angle for approach */

/* Only for SIDTYPE = CUSTOMDEPART */
const QLatin1String DEPARTURE_CUSTOM_DISTANCE("departurecustomdistance"); /* Length or final leg in NM */

/* Separator character for transition waypoint lists */
const QLatin1Char PROPERTY_LIST_SEP('#');

/* Copies all related properties and deletes the ones in "to" that do not exist in "from". */
void copySidProcedureProperties(QHash<QString, QString>& to, const QHash<QString, QString>& from);
void copyArrivalProcedureProperties(QHash<QString, QString>& to, const QHash<QString, QString>& from);
void copyStarProcedureProperties(QHash<QString, QString>& to, const QHash<QString, QString>& from);

/* Aircraft performance */
const QLatin1String AIRCRAFT_PERF_NAME("aircraftperfname");
const QLatin1String AIRCRAFT_PERF_TYPE("aircraftperftype");
const QLatin1String AIRCRAFT_PERF_FILE("aircraftperffile");

/* Source database navigation data */
const QLatin1String NAVDATA("navdata");
const QLatin1String NAVDATA_CYCLE("navdatacycle");

/* Source database simulator */
const QLatin1String SIMDATA("simdata");
const QLatin1String SIMDATA_CYCLE("simdatacycle");

/* Set if any of the airports to use DEP/DES instead of ADES/ADEP for loading in X-Plane FMS.
 * Read by FlightplanIO::saveFmsInternal() */
const QLatin1String AIRPORT_DEPARTURE_NO_AIRPORT("departnoapt");
const QLatin1String AIRPORT_DESTINATION_NO_AIRPORT("destnoapt");

/* Invalid/not set heading for parking position */
const float INVALID_HEADING = std::numeric_limits<float>::max();

/* Invalid position altitude for parking and departure */
const float INVALID_ALTITUDE = std::numeric_limits<float>::max();

} // namespace pln
} // namespace fs
} // namespace atools

#endif // ATOOLS_FLIGHTPLANCONSTANTS_H
`



![Screenshot (68)](https://github.com/albar965/littlenavmap/assets/117526105/341ed8d3-2a3e-4c94-b5e4-ea9ec2031964)

**Reply by albar965** [reply]
Again: The comment is broken. See my code above.

Current release branch is correct: https://github.com/albar965/atools/blob/release/4.0/src/fs/pln/flightplanconstants.h

Either check out the latest or fix it. :man_shrugging:

**Reply by Mohammad-Imran01** [reply]
I had absolutely no idea, it might be colliiding with any enums that i created. I also tried new versions as you said nothing worked then i had to go back to my 7/9days older codes. 
Image i shared above has correct asterisk, but text i pasted has not!!!!? Anyway leave it.

**Comment by Mohammad-Imran01** [comment]
I've been stuck on a problem for weeks, even after asking about it before [here](https://github.com/albar965/littlenavmap/discussions/1092#discussion-6252578). I made my own class called "PageInfoController" similar to "infoController" and connected most signals and slots the same way. At first, everything was fine; I got data when I changed the airport in the searchbase table.

But the problem is still there. When I keep changing airports (around 25 to 30 times) in the searchbase table, the app crashes with "Cn::Process::NotifyOutOfMemory()" or "sometimes image creation too many threads" errors. Even after completely removing infocontroller, the issue remains exactly same. Note that "PageInfoController" sends data to QAbstractListModel classes which update in qml pages using ListView. and I am using 32bit

**Reply by albar965** [reply]
I'm afraid I cannot help much with this. I'd need to dig into the code and debug into it but this takes an enormous amount of time.

**Reply by Mohammad-Imran01** [reply]
Okay Thanks alot............👍👍👍👍👍👍

**Comment by Mohammad-Imran01** [comment]
Hey,
I need to add 'Track mode' in which map moves and rotates plane stays stable. Can you please tell me where should i start or anything that might be helpful.
Thanks.

**Reply by albar965** [reply]
You can try to use https://doc.qt.io/qt-5/qpainter.html#setTransform to rotate the whole image before painting. But this will rotate the labels too and also destroy the projection. 

Anyway, huge effort. Think if you really need this. 
I get requests for this in LNM from time to time but refuse it since it is too much effort and would require to rewrite the drawing code.

**Reply by Mohammad-Imran01** [reply]
Thanks for your response.

**Comment by Mohammad-Imran01** [comment]
Hey, I need your help with writing my own queries. Like in search window(searchbaseTable), we write 'Ab' it gives rowsof all ident starting with same char. '  I did some additon in waypointquery.cpp. 

`static const QString whereLikeIdentRegion("ident LIKE :ident");
  waypointByLikeIdentQuery = new SqlQuery(dbNav);
  waypointByLikeIdentQuery->prepare("SELECT * FROM " + table + " WHERE " + whereLikeIdent + " LIMIT 1");`


Then using it 
` void FlightInformation::searchDirectTo(const QString &directToInput)
{
    if (waypointQuery != nullptr && root) {
        QList<MapWaypoint> wp;
        waypointQuery->getWaypointByLikeIdent(wp, directToInput + "%");
        const int wpLen = wp.length();
        if (wpLen > 0) {
            const MapWaypoint mapWp = wp.first(); // Safer to use first() with check for list emptiness
            root->setProperty("varDTAddressAirport", mapWp.region);
            qWarning() << Q_FUNC_INFO << "Waypoint address: " << mapWp.region;
        } else {
            qWarning() << Q_FUNC_INFO << "No waypoints found matching the input.";
        }
    } else {
        qWarning() << Q_FUNC_INFO << "Waypoint query or root unavailable.";
    }
} `

For any input, output: " qWarning() << Q_FUNC_INFO << "No waypoints found matching the input.";"

It gives no result, can you please tell me how should i create a class to run all my queries like mentioned above. I am very confused with sqlModel, searchBaseTable, Airport query, Sqlhandler classes.

**Reply by albar965** [reply]
Hard to say without the full code of `getWaypointByLikeIdent()`. The query and the code look ok to me.

**Comment by Mohammad-Imran01** [comment]
Hey I hope you are doing well, I was facing an issue where lnm installed on remote pc with prepar3D v4 works fine. When i compile and connect to that simulator from my PC it connects but aircrafts doesnt show up no matter what i do, So i compiled newer versions [lnm 3.0 release and atools 4.0 release] still same thing. The aircraft doesnt shows in compiled version on my PC, without any code changes,

![noAircraftIssue](https://github.com/albar965/littlenavmap/assets/117526105/ba3ffe1c-ffb5-4c41-a60e-a5db4b8df3d5)

**Reply by albar965** [reply]
This message indicates that you do not "sit in the cockpit" in the simulator. LNM needs updates from the sim to show the aircraft position.

**Reply by Mohammad-Imran01** [reply]
Okay Thanks alot.. i will check it.

**Comment by Mohammad-Imran01** [comment]
Hey, There is a problem which isnt going, and i cannot solve it i asked this problem last time too (https://github.com/albar965/littlenavmap/discussions/1092#discussioncomment-8955247) , there is a exact same 73 error, if i include any flightplan related file, i havent changed anything in them, if i dont include (flightplan related file )everything works fine, i thought it was beacuse my enum there is a collision i removed all of them nothing worked.I have tried the latest stable version too but its really 🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️🤦‍♂️.
Havent you been through this problem ever????
![unknownerror](https://github.com/user-attachments/assets/3ad4cc7a-76c2-43ba-b972-f458323c8332)

**Reply by Mohammad-Imran01** [reply]
flightplancontroller is  simple class(qabstractlistmodel) which includes route.h, and route.h includes flightplan, so i cannot inlcude flightplancontroller anywhere

**Reply by albar965** [reply]
Strange. This happens in the middle of the enum declaration (`ALTERNATE`). I'd guess a conflict with other enums but this one has its own namespace. Maybe include blocker `#ifdef' duplicated or forgotten?
Have no idea and this is not so easy to investigate offline.

**Reply by Mohammad-Imran01** [reply]
Thanks alot for your response, i'll look into it.

**Comment by Mohammad-Imran01** [comment]
Hi,
After installing the simulator on same pc, the LNM compiled is not connecting, am I forgetting something? Installed LNM easily connects everytime. I installed the 32bit simconnect as you said once but, only the error(simconnect not found) in the compiled version gone, but still not connecting. do i need to add some variable?

**Reply by Mohammad-Imran01** [reply]
installed version is 3.0.6 32bit and compiling version is 3.0.7 32bit, and it also works fine with simulator running on another pc

**Reply by albar965** [reply]
LNM loads the SimConnect.dll dynamically from the application/.exe folder. This is for 32-bit and 64-bit versions. The DLL is copied to the deploy/... folder when running `make deploy`: https://github.com/albar965/littlenavmap/blob/release/3.0/littlenavmap.pro#L989-L995

Path to the library can be defined by the environment variable `ATOOLS_SIMCONNECT_PATH_WIN32`.

**Reply by Mohammad-Imran01** [reply]
Thank you, but if simconnect is there, why it isnt connecting.
![disconnected01](https://github.com/user-attachments/assets/0b280270-b4d7-442c-acc9-f10532c4af93)

**Reply by Mohammad-Imran01** [reply]
![disconnected02](https://github.com/user-attachments/assets/85f6eee1-610c-4340-8a7c-2da124f3f230)

**Reply by Mohammad-Imran01** [reply]
Maybe different version problem?

**Reply by albar965** [reply]
For FSX/P3D and LNM 32-bit builds you have to use the version from the FSX SP2 SDK usually installed in `C:\Program Files (x86)\Microsoft Games\Microsoft Flight Simulator X SDK\SDK\Core Utilities Kit\SimConnect SDK\`.

For MSFS use the one that comes with their developer tools (`C:\MSFS SDK\SimConnect SDK\`).

Alex

**Comment by Mohammad-Imran01** [comment]
Hey i have one problem, and i really need your help.
I am receiving around 2kb/11ms data through QUdpSocket, which is causing UI issues, i tried alot to move the socket to another thread but still its blocking the main thread and overusing the CPU. I am sure issue is here 'while (udpSocket->hasPendingDatagrams()'

i asked on stackoverflow also, but no good reply.
https://stackoverflow.com/questions/78972335/cannot-move-udpsocket-to-another-thread-which-is-causing-ui-rendering-issues
https://stackoverflow.com/questions/78968765/qt-application-becomes-unresponsive-on-handling-high-frequency-data-coming-from?noredirect=1#comment139236768_78968765

```
#include "udphandler.h"
#include "custom/data/DataFile.h"
#include <QDebug>
#include <QtConcurrent/QtConcurrent>
#include <QTimer>

UdpSocketHandler::UdpSocketHandler(QObject *parent)
    : QObject(parent)
{
    udpSocket = new QUdpSocket(this);
    udpSocket->setSocketOption(QAbstractSocket::ReceiveBufferSizeSocketOption, 20 * 1024); //20kb
    udpSocket->setSocketOption(QAbstractSocket::LowDelayOption, 1);

    connect(udpSocket, &QUdpSocket::readyRead, this, &UdpSocketHandler::onReadyRead, Qt::DirectConnection);

}

UdpSocketHandler::~UdpSocketHandler() {
    if (udpSocket) {
        udpSocket->close();
        delete udpSocket;
    }
}

void UdpSocketHandler::close() {
    udpSocket->close();
}


void UdpSocketHandler::bindSocket(const QHostAddress &m_address, quint16 m_port)
{
    if (udpSocket && udpSocket->bind(m_address, m_port, QUdpSocket::ShareAddress | QUdpSocket::ReuseAddressHint)) {
        qWarning() << Q_FUNC_INFO << "Socket bound to Addr: " << m_address.toString() << " PORT: " << m_port;
    } else {
        qWarning() << Q_FUNC_INFO << "Failed to bind UDP socket. Addr: " << m_address.toString() << " PORT: " << m_port;
    }
}


void UdpSocketHandler::onReadyRead()
{
    while (udpSocket->hasPendingDatagrams()) {
        QByteArray datagram;
        datagram.resize(udpSocket->pendingDatagramSize());

        udpSocket->readDatagram(datagram.data(), datagram.size());

        aerx::readDataFromStruct(datagram);
    }
}

#include "netorkthread.h"



NetworkThread::NetworkThread(const QHostAddress &address, quint16 port, QObject *parent)
    : QThread(parent)
    , m_address(address)
    , m_port(port)
{
    setObjectName(QString("Network Thread"));
}

NetworkThread::~NetworkThread()
{
    requestInterruption();
    quit();
    wait();
}

void NetworkThread::init()
{
    qWarning() << "UDP Handler initialized in thread. " << QThread::objectName()
               << "\nAdd: " << m_address << ", Port: " << m_port;
    udpHandler = new UdpSocketHandler(nullptr);
    udpHandler->moveToThread(this);
    udpHandler->bindSocket(m_address, m_port);
}

void NetworkThread::deInit()
{
    if (udpHandler) {
        udpHandler->close();
        udpHandler->deleteLater();
    }
}

void NetworkThread::run() {
    init();
    exec();
    deInit();
}

// in main.cpp
NetworkThread networkThread(QHostAddress(offsetSenderIP), offsetSenderPORT);
networkThread.start();
QObject::connect(&app, &QCoreApplication::aboutToQuit, [&]() {
  networkThread.requestInterruption();
  networkThread.quit();
  networkThread.wait();
});
```

**Reply by albar965** [reply]
Not really sure since I just looked briefly over the code but your socket thread has no own event queue. That's why you have to use polling. I also suppose your code is not really running in the thread. The socket created in the constructor is still owned by the main thread.

Create a class derived from QObject which has all the logic and move this to a thread object (`moveToThread()`). Connect needed signals into this QObject derived and remove the polling.
Then use a queued connection to copy data from thread to the main thread.

[DataExchange](https://github.com/albar965/atools/blob/release/4.0/src/gui/dataexchange.h) uses this technique. The `DataExchangeFetcher` is the one derived from QObject.

**Reply by Mohammad-Imran01** [reply]
Thanks a lot.

**Comment by Mohammad-Imran01** [comment]
Hi
Can you tell me how to reduce the details, like roads and buildings from openTopoMap and cartodark, I tried to recompile after commenting respective lines in their .dgml but nothing changes, do i need to recompile the marble part?

**Reply by albar965** [reply]
You cannot change the details in these maps since they use pre-rendered map image tiles downloaded from the server by LNM. No DGML setting can change this.

You can try to use another map theme or use a MapBox user map which allows to change the details: https://www.littlenavmap.org/manuals/littlenavmap/release/latest/en/OPTIONS.html#map-web-service-configuration

**Comment by Mohammad-Imran01** [comment]
Hi hope you are doing well.
I am facing issues with connecting the compiled version lnm32 with p3d, Actually now i am trying to remove the simconnect part, i will send the data from a host applicaion via tcp/udp IPC.

Can you tell me if this is the only data, i will take care of related signals and ConnectDialog class and i will send data instead of reading it using simconnect.


struct SimData
{
  SimDataAircraft aircraft;

  float magVarDeg;
  float planeTrackMagneticDeg;
  float planeTrackTrueDeg;

  float planeAboveGroundFt;
  float groundAltitudeFt;

  qint32 autopilotAvailable;
  float altitudeAutopilotFt;

  float ambientTemperatureC;
  float totalAirTemperatureC;
  float ambientWindVelocityKts;
  float ambientWindDirectionDegT;

  qint32 ambientPrecipStateFlags;
  qint32 ambientIsInCloud;
  float ambientVisibilityMeter;
  float seaLevelPressureMbar;
  float pitotIcePercent;
  float structuralIcePercent;

  float airplaneTotalWeightLbs;
  float airplaneMaxGrossWeightLbs;
  float airplaneEmptyWeightLbs;
  float fuelTotalQuantityGallons;
  float fuelTotalWeightLbs;

  float fuelFlowPph1;
  float fuelFlowPph2;
  float fuelFlowPph3;
  float fuelFlowPph4;

  float fuelFlowGph1;
  float fuelFlowGph2;
  float fuelFlowGph3;
  float fuelFlowGph4;

  float localTimeSeconds;
  qint32 localYear;
  qint32 localMonth;
  qint32 localDay;
  float zuluTimeSeconds;
  qint32 zuluYear;
  qint32 zuluMonth;
  qint32 zuluDay;
  qint32 timeZoneOffsetSeconds;
};
struct SimDataAircraft
{
  char aircraftTitle[256];
  char aircraftAtcType[256];
  char aircraftAtcModel[256];
  char aircraftAtcId[256];
  char aircraftAtcAirline[256];
  char aircraftAtcFlightNumber[256];
  char category[256]; // "Airplane", "Helicopter", "Boat", "GroundVehicle", "ControlTower", "SimpleObject", "Viewer"

  qint32 userSim;
  qint32 modelRadius;
  qint32 wingSpan;

  float altitudeFt; // Actual altitude
  double latitudeDeg;
  double longitudeDeg;

  float groundVelocityKts;
  float indicatedAltitudeFt;

  float planeHeadingMagneticDeg;
  float planeHeadingTrueDeg;
  qint32 isSimOnGround;

  float airspeedTrueKts;
  float airspeedIndicatedKts;
  float airspeedMach;
  float verticalSpeedFps;

  qint32 transponderCode;

  qint32 numEngines;
  qint32 engineType; // 0 = Piston 1 = Jet 2 = None 3 = Helo(Bell) turbine 4 = Unsupported 5 = Turboprop

  char aiFrom[32];
  char aiTo[32];
};

**Reply by albar965** [reply]
Yes. This is the complete "raw" SimConnect structure which covers the user aircraft ,environment and all AI (aircraft and ships).
This is all copied into the class SimConnectData and its aggregates which is then serialized and moved around in the program.

Alex
